client/state/plugins/installed/selectors-ts.ts

```diff
@@ -95,3 +95,55 @@ export const getAllPluginsIndexedBySiteId = createSelector(
 	},
 	( state ) => [ getAllPluginsIndexedByPluginSlug( state ), getSiteIdsThatHavePlugins( state ) ]
 ) as { ( state: AppState ): { [ siteId: number ]: { [ pluginSlug: string ]: Plugin } } };
+
+export function getPluginsOnSites( state: AppState, plugins: Plugin[] ) {
+	return Object.values( plugins ).reduce(
+		( acc: { [ pluginSlug: string ]: Plugin }, plugin: Plugin ) => {
+			const siteIds = Object.keys( plugin.sites ).map( Number );
+			const pluginOnSites = getPluginOnSites( state, siteIds, plugin.slug );
+			if ( pluginOnSites ) {
+				acc[ plugin.slug ] = pluginOnSites;
+			}
+			return acc;
+		},
+		{}
+	);
+}
+
+export function getPluginOnSites( state: AppState, siteIds: number[], pluginSlug: string ) {
+	const plugin = getAllPluginsIndexedByPluginSlug( state )[ pluginSlug ];
+
+	if ( ! plugin ) {
+		return undefined;
+	}
+
+	for ( const siteId of siteIds ) {
+		if ( plugin.sites[ siteId ] ) {
+			return plugin;
+		}
+	}
+
+	return undefined;
+}
+
+export function getPluginOnSite( state: AppState, siteId: number, pluginSlug: string ) {
+	const plugin = getAllPluginsIndexedByPluginSlug( state )[ pluginSlug ];
+
+	const { sites, ...pluginWithoutSites } = plugin;
+
+	return plugin && plugin.sites[ siteId ]
+		? // Because this is a plugin for a single site context only the data for the selected
+		  // site is included. When I changed this file to TypeScript I found that some code
+		  // assumes it will be on the plugin object, while other code assumes it will be on the
+		  // sites object, so I included both.
+		  {
+				...pluginWithoutSites,
+				...plugin.sites[ siteId ],
```

Comment:
Dropping the plugin data with the plugin site data feels odd. Could be worth some comments here that explain what we're adding to this object.

Reply:
I changed the comment to try to clarify this.

It is odd behavior. This is basically a compromise between how the code worked before (putting site properties on the plugin object), and how it works now (site properties remain on site objects).

For this specific case only a single `siteId` is requested, and `getPlugins` was only adding data for the the sites that were requested. That means the object returned by this function always had only the data for a single site so there wasn't undefined behavior when the properties from different site objects were being put onto the same plugin.

There was some component code calling this selector seeking data for only a single site, and assuming that data was on the plugin object itself. I don't remember exactly why, but when I tried to refactor the component code it was difficult and so I made this compromise here.

I do think this could be worth addressing in a future PR, but I'm not sure if the dev time would be worth it in the end. 

If I had the time I would consider getting rid of all of the selectors in this PR but it would require a lot of changes in the component code. 

First, it is quite confusing to have four selectors all named in the form `/getPlugins?OnSites?/`. My head spins every time I try to distinguish them from one another and the differences in their behavior.

`getPluginOnSites` should be easily replaceable by having components call `getAllPluginsIndexedByPluginSlug` or `getAllPluginsIndexedBySiteId` directly.

`getPluginOnSite` should also be replaceable by having component code call `getAllPluginsIndexedByPluginSlug` or `getAllPluginsIndexedBySiteId` directly, but would require some refactoring so that component code doesn't assume that site data is on the plugin object, and also not assume that the plugin object has only data for the site requested instead of all the sites the plugin is on.

`getPluginsOnSite` is simple logic that doesn't really need to be in a selector, it is essentially an overload of `getPluginOnSite` where the final parameter is an array instead of a string.

`getPluginsOnSites` is sort of self-reflexive where it looks at each plugin object to see what sites it is on, and then gets the plugin for those sites. That should always be that plugin as far as I can tell: a plugin retrieved by its own sites should be itself every time, unless the component code was somehow passing in an incomplete `Plugin` object and getting back a completed one (which shouldn't be possible, and especially so now that the selectors are all in TypeScript). I'm not fully sure why this logic is needed, and I think instead a component could just call `getAllPluginsIndexedByPluginSlug` and then filter that down to only the plugins it needs, but such a change would have been more changes to component code so I didn't do it.